custom/enikshay/integrations/ninetyninedots/api_spec.py

```diff
@@ -81,45 +119,77 @@ def get_by_sector(self, prop, sector):
 class DotsApiParams(StrictJsonObject):
     api_params = jsonobject.ListProperty(DotsApiParam)
 
+    def get_param(self, param, sector):
+        try:
+            return next(
+                p for p in self.api_params
+                if p.api_param_name == param
+                and (getattr(p.case_properties, sector) or getattr(p.case_properties, 'both')
+                     or getattr(p.case_property, sector) or getattr(p.case_property, 'both'))
+            )
+        except StopIteration:
+            raise KeyError("{} not in spec".format(param))
+
+    @property
+    def required_params(self):
+        return [param.api_param_name for param in self.api_params if param.required_]
+
+    def params_with_choices(self, sector):
+        return [param.api_param_name for param in self.api_params if param.get_by_sector('choices', sector)]
+
     def params_by_case_type(self, sector, case_type):
         return [param for param in self.api_params
                 if param.get_by_sector('case_type', sector) == case_type]
 
-    def case_properties_by_case_type(self, sector, case_type):
+    def case_properties_by_case_type(self, sector, case_type, direction=DIRECTION_BOTH):
         params = self.params_by_case_type(sector, case_type)
         case_properties = []
         for param in params:
-            if param.get_by_sector("case_property", sector):
-                case_properties.append(param.get_by_sector("case_property", sector))
-            if param.get_by_sector("case_properties", sector):
-                case_properties += param.get_by_sector("case_properties", sector)
+            if param.direction & direction:
```

Comment:
`param.direction & direction` what does this ensure?

Reply:
this is a bitwise AND. Since direction is always a power of 2, it allows me to check if the direction matches easily (with "outbound" = 1, "inbound" = 2 and "both" = (1+2))